fdmi/plugin_dock_fom.c

```diff
@@ -301,25 +301,19 @@ static int pdock_fom_tick__feed_plugin_with_rec(struct m0_fom *fom)
 				 fids[pd_fom->pf_pos]);
 	if (rc == 0) {
 		/*
-		 * Plugin accepted the record,
-		 * so increment fdmi record refc
+		 * Plugin accepted the record. We already have a ref on
+		 * this record. No need to take extra ref.
+		 */
+		/*
+		 * If the plugin wants to do more work after returning
+		 * from the pcb->po_fdmi_rec(), it should increase fdmi
+		 * record refc, and then release it when done.
+		 * pseudo code:
+		 * struct m0_fdmi_record_reg *rreg;
+		 * rreg = m0_fdmi__pdock_record_reg_find(&frec->fr_rec_id);
+		 * if (rreg != NULL)
+		 *	m0_ref_get(&rreg->frr_ref);
 		 */
-		struct m0_fdmi_record_reg *rreg;
-
-		rreg = m0_fdmi__pdock_record_reg_find(&frec->fr_rec_id);
-
-		if (rreg == NULL) {
-			/* critical error,
-			   have to finish with no registration */
-			m0_fom_phase_set(fom,
-					 FDMI_PLG_DOCK_FOM_FINISH_WITH_REC);
-			M0_LOG(M0_ERROR,
-			       "record not registered in plugin dock");
-			M0_LEAVE();
-			return M0_FSO_AGAIN;
-		}
-
-		m0_ref_get(&rreg->frr_ref);
```

Comment:
This extra ref was balanced somewhere. We need to take extra care that now it will not go out of refc in some scenarios.

Reply:
When the release FOP is created, the ref count is set to 1.
And this ref count is dropped in pdock_fom_tick__finish_with_rec().

If an extra ref count is added in user provided "pcb->po_fdmi_rec()", the user plugin should drop the refcount when he/she finishes using it.


Here, this was a bug. Getting an extra ref count here causing unexpected result: one more refcount is not released at all.

Reply:
The reason why we did not see it before is probably that it affected the cleanup of the service, caused memory leaks, and inability to release the fol rec ut it did not crash in a "loud" way. Good catch then!

Reply:
Yes. The release fop is never sent, because the last ref is never released.
With this fix, the last ref is released, and then I fixed some issues in fdmi_sample_plugin about the fdmi server stop.
The bug in fdmi_sample_plugin was not revealed either, because the ref count is not dropped and release fop is not sent.